lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -1054,31 +1054,82 @@ def _stop_opensearch(self, *, restart=False) -> None:
-                logger.debug("Failed to get online nodes, voting and alloc exclusions not added")
+        if not self.opensearch.is_started():
+            self.status.set(WaitingStatus(ServiceStopped))
+            # Nothing to do, the service is not yet up
+            return
 
-        # TODO: should block until all shards move addressed in PR DPE-2234
+        # 1. stop the service
+        try:
+            self.opensearch.stop()
+            self.peers_data.delete(Scope.UNIT, "started")
+            self.status.set(WaitingStatus(ServiceStopped))
+        except OpenSearchStopError as e:
+            logger.exception(e)
+            # Now, we need to decide if the service is still functional or not
+            if not self.opensearch.is_node_up():
+                # We are still running, there was an error
+                raise
 
-        # 2. stop the service
-        self.opensearch.stop()
-        self.peers_data.delete(Scope.UNIT, "started")
-        self.status.set(WaitingStatus(ServiceStopped))
+        # 2. Check if we have enough voting units
+        self._settle_voting_exclusions(exclude_this_unit=True)
+
+    def _settle_voting_exclusions(self, exclude_this_unit: bool = False):
+        """Settle the exclusions for all voting units, following:
+
+        1) Any unit in /cat/_nodes that can vote should not be in the exclusion list
+        2) If exclude_this_unit: we remove it from the node list and we set it as excluded
+        2) Check if the unit count <= 2, in this case we need to manually manage exclusions
+
+        Args:
+            exclude_this_unit: Exclude the current unit from the considerations, useful if this
+                unit is being stopped and we don't want it to show up in any of the lists.
+        """
+        hosts = self.alt_hosts
+        if (
+            self.opensearch_peer_cm.deployment_desc().typ != DeploymentType.MAIN_ORCHESTRATOR
+            and (peer_cm_rel_data := self.opensearch_peer_cm.rel_data()) is not None
+        ):
+            # Also consider peer-relation units
+            hosts.extend([node.ip for node in peer_cm_rel_data.cm_nodes])
+
+        nodes = ClusterTopology.nodes(self.opensearch, use_localhost=False, hosts=hosts)
+        if exclude_this_unit:
+            self.opensearch_exclusions.delete_current()
+            if self.opensearch.current().name in nodes:
+                # It is possible the current node will still be on the list
+                nodes.remove(self.opensearch.current().name)
+
+        cms = ClusterTopology.get_cluster_managers_names(nodes)
+
+        # Review the exclusion list
+        self.opensearch_exclusions.delete_voting(cms)
 
-        # 3. Remove the exclusions
-        # TODO: we should probably NOT have any exclusion on restart
-        # https://chat.canonical.com/canonical/pl/bgndmrfxr7fbpgmwpdk3hin93c
-        self.opensearch_exclusions.delete_current()
+        leader = ClusterTopology.elected_manager_name(
```

Comment:
Should we maybe not call it "leader" as this may be confusing with the juju leader?